src/xcode/ENA/ENA/Source/Services/Event Registration/TraceWarningDownload/TraceWarningPackageDownload.swift

```diff
@@ -0,0 +1,400 @@
+		removePackagesFromTraceWarningMetadataPackagesTable(packagesToDelete)
+	}
+	
+	/// Filters out the packages to be downloaded by subtracting the actual packages from the available packages. Not private for testing purposes.
+	func determinePackagesToDownload(
+		availables availablePackagesOnCDN: [Int],
+		earliest earliestRelevantPackage: Int
+	) -> Set<Int> {
+		// Get all packages that are earlier then the earliestRelevantPackage
+		let earlierPackages = Set(availablePackagesOnCDN.filter { return $0 >= earliestRelevantPackage })
+		// Now filter out all entries that are not in our metadata database
+		let metadataIds = Set(eventStore.traceWarningPackageMetadatasPublisher.value.map { $0.id })
+		return earlierPackages.subtracting(metadataIds)
+	}
+	
+	// MARK: - Private
+	
+	private let client: Client
+	private let store: Store
+	private let eventStore: EventStoringProviding
+	private let countries: [Country.ID]
+	private let matcher: TraceWarningMatching
+	private let signatureVerifier: SignatureVerification
+	
+	private var subscriptions: Set<AnyCancellable> = []
+	private var status: TraceWarningDownloadStatus = .idle {
+		didSet {
+			statusDidChange?(status)
+		}
+	}
+	
+	private func checkForDownloadTraceWarningPackages(
+		with appConfig: SAP_Internal_V2_ApplicationConfigurationIOS,
+		countries: [Country.ID],
+		completion: @escaping (Result<TraceWarningSuccess, TraceWarningError>) -> Void
+	) {
+		let countriesDG = DispatchGroup()
+		var errors = [TraceWarningError]()
+		var successes = [TraceWarningSuccess]()
+		
+		// Download packages for each country
+		countries.forEach { country in
+			Log.info("TraceWarningPackageDownload: Start processsing package download for country: \(country).", log: .checkin)
+			
+			// Check if we did not discover in the same hour before.
+			if shouldStartPackageDownload(for: country) {
+				countriesDG.enter()
+				
+				// Go now for the real download
+				downloadTraceWarningPackages(with: appConfig, for: country, completion: { result in
+					switch result {
+					case let .success(success):
+						Log.info("TraceWarningPackageDownload: Succeded downloading packages for country id: \(country).", log: .checkin)
+						successes.append(success)
+					case let .failure(error):
+						Log.info("TraceWarningPackageDownload: Failed downloading packages for country id: \(country).", log: .checkin)
+						errors.append(error)
+					}
+					
+					countriesDG.leave()
+				})
+			}
+		}
+		
+		countriesDG.notify(queue: .main) { [weak self] in
+			if let error = errors.first {
+				Log.error("TraceWarningPackageDownload: Failed downloading packages for all countries with errors: \(errors).", log: .checkin)
+				self?.store.wasRecentTraceWarningDownloadSuccessful = false
+				completion(.failure(error))
+			} else {
+				Log.info("TraceWarningPackageDownload: Completed downloading packages for all countries.", log: .checkin)
+				self?.store.lastTraceWarningPackageDownloadDate = Date()
```

Comment:
Is lastTraceWarningPackageDownloadDate needs to be used somewhere? I only see usage in the tests.

Reply:
I think you will need it for the riskCalculation in the RiskProvider, like `lastKeyPackageDownloadDate` is used only there.

Reply:
After short discussion with Artur, I will remove the property. If we need it later, we can re-add it.